dbms/src/Common/CachingAllocator.h

```diff
@@ -0,0 +1,875 @@
+     */
+    TRegionsList all_regions;
+
+    /**
+     * Represented by a size multimap.
+     * Used to find free region with at least requested size.
+     */
+    TFreeRegionsMap free_regions;
+
+    /**
+     * Represented by a keymap (boost multiset).
+     */
+    TAllocatedRegionsMap allocated_regions;
+
+    /**
+     * Represented by a LRU List (boost linked list).
+     * Used to find the next region for eviction.
+     * TODO Replace with exponentially-smoothed size-weighted LFU map.
+     */
+    TUnusedRegionsList unused_allocated_regions;
+
+    struct MemoryChunk;
+    std::list<MemoryChunk> chunks;
+
+    struct InsertionAttempt;
+    std::unordered_map<Key, std::shared_ptr<InsertionAttempt>, THash> insertion_attempts;
+
+    mutable std::mutex mutex;
+
+    pcg64 rng{randomSeed()};
+
+    /**
+     * Stats data
+     */
+
+    /// Max cache size.
+    const size_t max_total_size;
+
+    size_t total_chunks_size {0};
+    size_t total_allocated_size {0};
+    size_t total_size_in_use {0};
+
+    std::atomic<size_t> total_size_currently_initialized {0};
+
+    /// Value was in the cache.
+    std::atomic<size_t> hits {0};
+
+    /// Value we were waiting for was calculated by another thread.
+    /// @note Also summed in \c hits.
+    std::atomic<size_t> concurrent_hits {0};
+
+    /// Value was not found in the cache
+    std::atomic<size_t> misses {0};
+
+    size_t allocations {0};
+    size_t allocated_bytes {0};
+    size_t evictions {0};
+    size_t evicted_bytes {0};
+    size_t secondary_evictions {0};
+
+    /**
+     * Represents an attempt to insert a new entry into cache.
+     * Useful in 2 situation:
+     * <ul>
+     * <li>
+     * Value for a given key was not found in the cache, but it may be produced by another thread.
+     * We create this \c attempt and try to achieve a lock on its \c mutex.
+     * On success we may find the \c attempt already has a \c value (= it was calculated by other thread).
+     * </li>
+     * <li>
+     * Same as 1, but the \c attempt has no \c value. It simply means value is not present in cache, so
+     * we approve this attempt and insert it.
+     * </li>
+     * </ul>
+     *
+     * @see \c GetOrSet
+     */
+    struct InsertionAttempt
+    {
+        explicit InsertionAttempt(CachingAllocator & cache_) noexcept : cache(cache_) {}
+
+        CachingAllocator & cache;
+
+        /// Mutex that protects \c is_disposed, \c value and \c refcount
+        std::mutex mutex;
+
+        bool is_disposed{false};
+        MemoryRegionPtr value;
+        size_t refcount {0};
+    };
+
+    /**
+     * Responsible for removing a single \c InsertionAttempt from the \c insertion_attempts container.
+     * The disposal may be performed via an explicit \c dispose call or in the destructor.
+     *
+     * Among several concurrent threads the first successful one is responsible for removal.
+     * If they all fail, the last one is responsible.
+     */
+    struct InsertionAttemptDisposer
```

Comment:
The purpose of InsertionAttempt and InsertionAttemptDisposer is to support concurrent initialization of the cache entries, right? Looks like too much boilerplate. Maybe it would be simpler to keep some status field in the cache entry metadata, and wait on it until the entry becomes ready. Probably one condvar for entire cache would suffice, since the initializations are not so frequent, but if not, condvars are cheap.

Reply:
Ok, I will replace these structs with proper condvars.